
# Case Técnico – Analista de Dados (Databricks-First)

Este notebook foi preparado **para execução direta no Databricks (Free Edition ou superior)** usando **Pyspark, Python ou SQL**. <br> 
Você vai trabalhar com dois CSVs:
- `dim_clientes.csv`
- `fato_receita.csv`

> **Como preparar o ambiente (1 minuto):**
> 1. No Databricks, vá em **Workspace → Files** (ícone de pasta).
> 2. Crie uma pasta `banking_case` (opcional) e **faça upload** dos dois CSVs para dentro dela.
> 3. Anexe um **Compute** (serverless/cluster) ao notebook (canto superior direito).
> 4. Execute a seção **Setup** abaixo.
> 5. Inclua seus comentários/assunções no próprio notebook.


## 📤 Exportar e Enviar

Quando terminar seu case:
- **File → Export**: baixe como **HTML** ou **Jupyter (.ipynb)**;
- Ou publique num repositório git e compartilhe o link.

Nos retorne o notebook com suas respostas em anexo.



## Setup (Databricks)

O código abaixo:
- Define o seu usuário
- Recebe o input da pasta onde você subiu os arquivos
- Cria as tabelas no notebook para serem trabalhadas via **python**/sql/pyspark


In [0]:
import pandas as pd
import numpy as np
from pyspark.sql import functions as F
from pyspark.sql.functions import *
from pyspark.sql import Window, SQLContext

In [0]:
pasta = "case" ##nome da pasta onde você subiu o notebook
usuario = spark.sql("SELECT current_user()").collect()[0][0]
print(usuario)
## --- ## Bases em Python utilizando Pandas
clientes = pd.read_csv("/Workspace/Users/{0}/{1}/dim_clientes.csv".format(usuario, pasta))
receita = pd.read_csv("/Workspace/Users/{0}/{1}/fato_receita.csv".format(usuario, pasta))



## --- ## Bases em Pyspark utilizando Pyspark
# spark_dim_clientes = spark.createDataFrame(clientes)
# spark_fato_receita = spark.createDataFrame(receita)

## --- ## Bases para SQL
# spark_dim_clientes.createOrReplaceTempView("clientes")
# spark_fato_receita.createOrReplaceTempView("receita")

lucas.silvestrini@live.com



## ✅ Pergunta 1

O banco está revisando sua estratégia comercial e precisa entender melhor a base de clientes e a performance de cada produto. 

Você foi chamado para apoiar o time executivo na construção de um resumo que ajude a responder perguntas simples, mas fundamentais.

O Head de Negócio perguntou: <br> 
<b>Se fossemos focar na melhoria de um produto apenas para o próximo ano, 
qual seria e que métricas você utilizaria para avaliar esta decisão?</b>

- Qual a relevância de cada produto na nossa base? 
- Qual a receita média por cliente de cada produto?
- Algum produto piorou ou melhorou sua performance?

#### ⏳ Desenvolvimento 01:

In [0]:
# a1 Qual produto para focar melhoria?
# a2 Quais as métricas utilizadas?
# a3 Qual a relevância de cada produto na nossa base? 
# a4 Qual a receita média por cliente de cada produto?
# a5 Algum produto piorou ou melhorou sua performance?

# 1o: entender as bases e variáveis
# dim_clientes contém as características dos clientes, idade, renda, perfil digital, regiao, tempo de relacionamento.
# fato_receita contém a receita mensal por produto por cliente

# Apaguei a ultima linha em branco manual ao verificar os csv's, mas poderia ter feito com dropna()

print('Dimensão clientes:', clientes.shape)
print('Clientes únicos na dimensão:', clientes['cliente_id'].nunique())

print('Fato receita:', receita.shape)
print('Clientes únicos na receita:', receita['cliente_id'].nunique())

# merge left entre as bases, trazendo as informacoes de receita e anomes para a clientes.
db1 = receita.merge(clientes, on='cliente_id', how='left')
print('db1 integrada:', db1.shape)
print('Clientes únicos na db1 :', db1['cliente_id'].nunique())

#verificacoes iniciais periodo, produtos, regioes e perfil digital
print('\nPeríodo da receita:', receita['ano_mes'].min(), 'a', receita['ano_mes'].max())

print('\n Produtos:', sorted(receita['produto_id'].unique()))
# verificar qual a distribuiçao de produto_id da base
print(db1['produto_id'].describe())

print('\n Perfil digital:', sorted(clientes['perfil_digital'].unique()))
# verificar qual a distribuiçao de perfil_digital da base
print(clientes['perfil_digital'].describe())

print('\n renda_mensal:', clientes['renda_mensal'].min(), 'a', clientes['renda_mensal'].max())
# verificar qual a faixa de renda da base
print(clientes['renda_mensal'].describe())

print('\n Regiões:', sorted(clientes['regiao'].unique()))
# verificar qual a distribuiçao de regiões da base
print(clientes['regiao'].describe())

# Com essas informações já é possível entender: 
# 1) no periodo apresentado o banco tem 1000 clientes
# 2) o banco tem 5 produtos: cartao consorcio, credito, invest e seguro
# 2) o maior produto do banco é o cartão, com 8928 lançamentos na base de receita ao longo do ano de 2024.
# 3) o perfil digital da base é alto com 667, 66,7% da base de mil clientes
# 4) a renda varia de 2 a 16,5k, com a média em 3.6k e a 75% concentrada na faixa de 0 a 4,4k.
# 5) o banco divide os clientes por 5 regiões, sendo o sudeste o principal com 449 clientes 44,9% da base.


Dimensão clientes: (1000, 7)
Clientes únicos na dimensão: 1000
Fato receita: (14403, 4)
Clientes únicos na receita: 1000
db1 integrada: (14403, 10)
Clientes únicos na db1 : 1000

Período da receita: 2024-01 a 2024-12

 Produtos: ['CARTAO', 'CONSORCIO', 'CREDITO', 'INVEST', 'SEGURO']
count      14403
unique         5
top       CARTAO
freq        8928
Name: produto_id, dtype: object

 Perfil digital: ['ALTO', 'BAIXO', 'MEDIO']
count     1000
unique       3
top       ALTO
freq       667
Name: perfil_digital, dtype: object

 renda_mensal: 2000.0 a 16559.78
count     1000.000000
mean      3665.196420
std       1885.234182
min       2000.000000
25%       2195.110000
50%       3139.395000
75%       4494.107500
max      16559.780000
Name: renda_mensal, dtype: float64

 Regiões: ['CENTRO-OESTE', 'NORDESTE', 'NORTE', 'SUDESTE', 'SUL']
count        1000
unique          5
top       SUDESTE
freq          449
Name: regiao, dtype: object


In [0]:

display(db1)

# display no db1 para mostrar a tabela e permitir a visualizacao rapidas e entendimento dos dados macro

# na primeira visualizacao criei uma pivot de anomes por produto somando os clientes unicos, para entender a variacao da receita ao longo do tempo. Aqui é possivel perceber que temos poucos dados em jan, entao pode ser que os outros produtos ainda nao haviam começado nessa data, tendo apenas o produto cartao com os 214 clientes.
# mesmo o cartao sendo o melhor produto na qtd de clientes, é perceptivel o crescimento do credito, invest e seguros.

# na segunda visualizacao, criei outra pivot olhando para receita, cartoes mantem uma receita estavel, afinal está com clientes estaveis, porém credito e invest tem ganhos consideraveis. Em seguida alterei para receita média, assim é possivel entender o comportamento da receita por cliente, consorcio tem a maior média em dez24, seguido por invest e credito.

# na terceira verifiquei as outras variaveis alternando a coluna para regiao, renda, tempo de relacionamento e idade, para entender melhor o posicionamento da carteira, por fim deixei na visão anomes, produto e renda média.



cliente_id,produto_id,ano_mes,receita,idade,renda_mensal,perfil_digital,regiao,tempo_relacionamento_anos,dt_abertura_conta
100000,CARTAO,2024-04,630.6,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-05,589.99,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-06,967.19,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-07,385.5,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-08,723.17,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-09,297.97,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-10,98.89,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-11,577.1,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,CARTAO,2024-12,633.14,49,6583.23,MEDIO,SUL,1.2,2023-11-07
100000,SEGURO,2024-10,266.34,49,6583.23,MEDIO,SUL,1.2,2023-11-07


Databricks visualization. Run in Databricks to view.

Databricks visualization. Run in Databricks to view.

Databricks visualization. Run in Databricks to view.

In [0]:
# verificacao se temos churn na base:
# não tem churn em nenhum produto, isto é, todos os clientes que contratam um produto nessa base continuam gerando receita nesse produto, sem cancelamento.

receita["ano_mes"] = pd.to_datetime(receita["ano_mes"])

ultimo_mes_base = receita["ano_mes"].max()

print("Último mês da base:", ultimo_mes_base)

churn_produto = (
    receita
    .groupby(["cliente_id", "produto_id"])
    .agg(
        primeiro_mes=("ano_mes", "min"),
        ultimo_mes=("ano_mes", "max"),
        meses_ativos=("ano_mes", "nunique"),
        receita_total=("receita", "sum")
    )
    .reset_index()
)

churn_produto["churn_produto"] = churn_produto["ultimo_mes"] < ultimo_mes_base

resumo_churn_produto = (
    churn_produto
    .groupby("produto_id")
    .agg(
        clientes_produto=("cliente_id", "count"),
        clientes_com_churn=("churn_produto", "sum")
    )
    .reset_index()
)

resumo_churn_produto["taxa_churn_produto"] = (
    resumo_churn_produto["clientes_com_churn"] 
    / resumo_churn_produto["clientes_produto"]
)

resumo_churn_produto


Último mês da base: 2024-12-01 00:00:00


,produto_id,clientes_produto,clientes_com_churn,taxa_churn_produto
0,CARTAO,860,0,0.0
1,CONSORCIO,102,0,0.0
2,CREDITO,492,0,0.0
3,INVEST,386,0,0.0
4,SEGURO,301,0,0.0


In [0]:
# após entender o comportamento de cada produto ao longo do tempo de forma geral, criei uma tabela com os dados agregados por produto para responder a primeira questao.

#groupby por produto da db1 com a soma da receita, clientes unicos, media de receita por mês, 
db2 = (
    db1.groupby('produto_id')
        .agg(
            receita_total=('receita', 'sum'),
            clientes=('cliente_id', 'nunique'),
            receita_media_cliente_mes=('receita', 'mean')
    )
    .reset_index()
)

# criando as variaveis de receita media por ano, %receita e %pn
db2['receita_media_cliente_ano'] = db2['receita_total'] / db2['clientes']
db2['% Receita'] = 100 * db2['receita_total'] / db2['receita_total'].sum()
clientes_totais = clientes['cliente_id'].nunique()
db2['Penetração (%)'] = 100 * db2['clientes'] / clientes_totais

#renomeando para melhorar a visualizacao da tabela
db2 = db2.rename(columns={
    'produto_id': 'Produto',
    'clientes': 'Clientes Unicos',
    'receita_total': 'Receita Total',
    'receita_media_cliente_ano': 'Receita Média por cliente',
    'receita_media_cliente_mes': 'Receita Média por Cliente por mês'
})

#display da tabela final, ordenando pela receita.
display(db2[[
    'Produto',
    'Clientes Unicos',
    'Penetração (%)',
    'Receita Total',
    '% Receita',
    'Receita Média por cliente',
    'Receita Média por Cliente por mês'
]].sort_values('Receita Total', ascending=False))

# Com esses dados é possivel respondermos algumas questoes, 
# 1o o produto com maior receita e rentabilidade é o cartao, com 860 clientes, 5,17 mi de receita, 6k de receita por cliente anual e 579 mensal, contando com 86% de penetração, isto é, a cada 100 clientes 86 tem o produto cartao nessa base.

#  apenas olhando para esses dados, a resposta seria que cartoes é o melhor produto para investir, porém, acredito que os produtos de invest, seg e consorcio podem apresentar maior potencial de crescimento de receita visto que a receita media por cliente é bem maior. 
# Consorcio é o produto mais rentavel da base, com 1144 de receita media por cliente, ainda é um produto em desenvolvimento com base de clientes menor.

# a resposta quanto a qual produto focar melhorias fica será o cartao por ser o maior produto do banco, que gera a principal receita atual e tem maior número de clientes. Porém, tem o ponto de saturação, a base está estavel durante alguns meses, precisaria de mais variaveis de negócios/produtos para entender qual vale mais a pena investir.
# Investimentos e credito podem ser considerados como produtos para melhoria em formato de cross-sell com o cartão, dependendo de outras questoes como as complexidades de melhorias e custos agregados (operacoes, infra, risco, etc.)

Produto,Clientes Unicos,Penetração (%),Receita Total,% Receita,Receita Média por cliente,Receita Média por Cliente por mês
CARTAO,860,86.0,5170247.88,57.35762536624514,6011.916139534884,579.1048252688172
INVEST,386,38.6,1581645.81,17.54644070026123,4097.528005181347,945.3949850567842
CREDITO,492,49.2,1404513.14,15.58136870336836,2854.7015040650404,679.1649613152804
CONSORCIO,102,10.2,520523.96,5.774581603200497,5103.176078431373,1144.0087032967033
SEGURO,301,30.1,337124.18,3.7399836269247873,1120.0138870431892,263.5841907740422


In [0]:
# Receita e clientes únicos por produto e mês
receita_clientes_mes = (
db1.groupby(['produto_id', 'ano_mes'])
    .agg(
    receita_total=('receita', 'sum'),   
           receita_media=('receita', 'mean'),
        clientes_unicos=('cliente_id', 'nunique')
    )
    .reset_index()
    .sort_values(['produto_id', 'ano_mes'])
)

# Mudanças mês a mês por produto_id
receita_clientes_mes['delta_receita'] = receita_clientes_mes.groupby('produto_id')['receita_total'].diff()
receita_clientes_mes['delta_clientes'] = receita_clientes_mes.groupby('produto_id')['clientes_unicos'].diff()

display(receita_clientes_mes)

# Aqui, criei um gráfico de Bubble para visualizar a evolução da receita e clientes por produto e mês, visualmente facilita a percepcao do crescimento do invest e credito, e a estabilização do cartao.

produto_id,ano_mes,receita_total,receita_media,clientes_unicos,delta_receita,delta_clientes
CARTAO,2024-01,127819.88,597.2891588785047,214,null,null
CARTAO,2024-02,258123.34,617.51995215311,418,130303.45999999999,204.0
CARTAO,2024-03,410637.03,651.8048095238096,630,152513.69000000003,212.0
CARTAO,2024-04,580679.06,705.5638639125152,823,170042.03000000003,193.0
CARTAO,2024-05,523773.24,636.419489671932,823,-56905.820000000065,0.0
CARTAO,2024-06,546999.78,636.0462558139535,860,23226.540000000037,37.0
CARTAO,2024-07,496453.47,577.271476744186,860,-50546.310000000056,0.0
CARTAO,2024-08,457763.29,532.2828953488372,860,-38690.17999999999,0.0
CARTAO,2024-09,431917.59,502.22975581395355,860,-25845.699999999953,0.0
CARTAO,2024-10,435463.94,506.35341860465115,860,3546.3499999999767,0.0


Databricks visualization. Run in Databricks to view.


#### ⚪ Resposta 01: 
Após verificar os dados e fazer algumas análises, a resposta é que temos 3 caminhos para seguir:

1) Seguir com melhorias no maior produto da carteira, o cartão.

As métricas utilizadas para essa decisão, com base nos dados disponíveis são: 
- Maior base de clientes 860, representando 86% da base
- Maior receita anual 5.1 milhões
- Quantidade de clientes e receita está estabilizada nos ultimos 5 meses, necessário verificar se a aquisição diminuiu, se manteve mas aumentou o churn ou se os clientes estão com limite total tomado.
- A ideia de seguir com o cartão é manter a receita alta do produto entrando, para poder continuar garantindo o fluxo de receita e aumentar cross-sell para os produtos menores.



2) Escalar investimentos ou crédito
Estão em crescimento acelerado e processo de maturação, podem alcançar maior número de clientes e resultados do que cartões no futuro, caso recebam melhorias e campanhas específicas.

- São os maiores produtos além do cartão, tem melhor receita média por cliente por mês e podem agregar valor ao uso do principal produto com campanhas abrangendo ambos produtos.


3) Melhorias em consórcio, é o produto que apresenta a maior receita média por cliente por mês, ainda está em processo de crescimento e ao chegar no tamanho da carteira de cartão, gerará muito mais receita, porém, é necessário entender perfil da base e interesse de contratação do produto.

> A resposta final é que o banco deve aprofundar o estudo entre cartão, investimentos ou crédito, dependendo de outras variaveis (infra, produtos, marketing, budget, crédito, risco etc.)


## ✅ Pergunta 2

Além do perfil dos nossos produtos, nosso Head de negócios gostaria de um breve diagnóstico dos nossos clientes:
1. Existe algum perfil de renda que atendemos melhor?
2. O tempo de relacionamento implica em resultados melhores? Estamos rentabilizando nossos clientes à medida que eles permanecem na base?



#### ⏳ Desenvolvimento 02:

In [0]:
# verificar qual a faixa de renda da base
display(clientes['renda_mensal'].describe())
# a média de renda é 3.6k, o minimo 2k e até 75% da base em 4.5k, assim podemos inferir as faixas 0 a 3, 3 a 5, 5 a 8 e 8+

# Verificar a quantidade total de receita, clientes e share por região usando groupby, em seguida colocar a faixa de renda.
cliente_metricas = (
    db1.groupby('cliente_id')
    .agg(
        receita_total=('receita', 'sum'),
        produtos=('produto_id', 'nunique'),
        meses_com_receita=('ano_mes', 'nunique')
    )
    .merge(clientes, on='cliente_id', how='left')
)

# criar as faixas, pd.cut ou com if/elif.
cliente_metricas['faixa_renda'] = pd.cut(
    cliente_metricas['renda_mensal'],
    bins=[0, 3000, 5000, 8000, np.inf],
    labels=['Até R$3k', 'R$3k–5k', 'R$5k–8k', 'R$8k+'],
)


# agrupar por faixa de renda
renda_resumo = (
    cliente_metricas.groupby('faixa_renda', observed=True)
    .agg(
        clientes=('cliente_id', 'nunique'),
        renda_media=('renda_mensal', 'mean'),
        receita_total=('receita_total', 'sum'),
        receita_media_cliente=('receita_total', 'mean'),
        produtos_medios_cliente=('produtos', 'mean')
    )
    .reset_index()
)

clientes_total = clientes['cliente_id'].nunique()
renda_resumo['share_clientes'] = renda_resumo['clientes'] / clientes_total

receita_total = db1['receita'].sum()
renda_resumo['share_receita'] = renda_resumo['receita_total'] / receita_total

display(renda_resumo)



count     1000.000000
mean      3665.196420
std       1885.234182
min       2000.000000
25%       2195.110000
50%       3139.395000
75%       4494.107500
max      16559.780000
Name: renda_mensal, dtype: float64

faixa_renda,clientes,renda_media,receita_total,receita_media_cliente,produtos_medios_cliente,share_clientes,share_receita
Até R$3k,463,2287.238963282937,4176136.94,9019.734211663066,2.140388768898488,0.463,0.46329170987959933
R$3k–5k,353,3840.9586968838526,3215787.28,9109.878980169971,2.189801699716714,0.353,0.35675257036955926
R$5k–8k,144,5957.803263888889,1225368.63,8509.504374999999,2.0277777777777777,0.144,0.13593977783341607
R$8k+,40,9810.56725,396762.12,9919.053,2.125,0.04,0.04401594191742543


In [0]:
# como a maior parte dos clientes, 463 e 353 estão nas faixas até 5k,
#  talvez, olhar mais aprofundado essas faixas mais baixas: 0-2, 2-3, 3-4, 4-5, 5-8, 8+
cliente_metricas['faixa_renda2'] = pd.cut(
    cliente_metricas['renda_mensal'],
    bins=[0, 2000, 3000,4000,5000, 8000, np.inf],
    labels=['Até R$2k', 'R$2k–3k','R$3k–4k', 'R$4k–5k','R$5k–8k', 'R$8k+'],
)

# agrupar por faixa de renda
renda_resumo2 = (
    cliente_metricas.groupby('faixa_renda2', observed=True)
    .agg(
        clientes=('cliente_id', 'nunique'),
        renda_media=('renda_mensal', 'mean'),
        receita_total=('receita_total', 'sum'),
        receita_media_cliente=('receita_total', 'mean'),
        produtos_medios_cliente=('produtos', 'mean')
    )
    .reset_index()
)

clientes_total = clientes['cliente_id'].nunique()
renda_resumo2['share_clientes'] = renda_resumo2['clientes'] / clientes_total

receita_total = db1['receita'].sum()
renda_resumo2['share_receita'] = renda_resumo2['receita_total'] / receita_total

display(renda_resumo2)


faixa_renda2,clientes,renda_media,receita_total,receita_media_cliente,produtos_medios_cliente,share_clientes,share_receita
Até R$2k,196,2000.0,1758606.01,8972.479642857143,2.1377551020408165,0.196,0.19509599351822016
R$2k–3k,267,2498.0960299625467,2417530.93,9054.422958801499,2.142322097378277,0.267,0.2681957163613792
R$3k–4k,227,3465.364625550661,2045608.14,9011.48960352423,2.202643171806167,0.227,0.22693539664535683
R$4k–5k,126,4517.624206349206,1170179.14,9287.136031746031,2.1666666666666665,0.126,0.1298171737242024
R$5k–8k,144,5957.803263888889,1225368.63,8509.504374999999,2.0277777777777777,0.144,0.13593977783341607
R$8k+,40,9810.56725,396762.12,9919.053,2.125,0.04,0.04401594191742543


In [0]:

# verificar qual a faixa de tempo de relacionamento da base
display(clientes['tempo_relacionamento_anos'].describe())

# a média de tempo de relacionamento é 10 anos, 25% em 5 anos e 50% em 10, possivel fazer os cortes em 0-2 anos novos clientes, 2-5 clientes recentes, 5-10 clientes fidelizados, 10-15 clientes longo prazo e 15+ clientes fieis

# criar as faixas de tempo e agrupar para verificacao 
cliente_metricas['faixa_tempo'] = pd.cut(
    cliente_metricas['tempo_relacionamento_anos'],
    bins=[0, 2, 5, 10, 15, np.inf],
    labels=['0–2 anos', '2–5 anos', '5–10 anos', '10–15 anos', '15+ anos'],
    right=False
)

tempo_resumo = (
    cliente_metricas.groupby('faixa_tempo', observed=True)
    .agg(
        clientes=('cliente_id', 'nunique'),
        tempo_medio=('tempo_relacionamento_anos', 'mean'),
        receita_total=('receita_total', 'sum'),
        receita_media_cliente=('receita_total', 'mean'),
        produtos_medios_cliente=('produtos', 'mean'),
        meses_com_receita_medios=('meses_com_receita', 'mean')
    )
    .reset_index()
)
tempo_resumo['share_receita'] = tempo_resumo['receita_total'] / receita_total
display(tempo_resumo)


count    1000.000000
mean       10.569400
std         6.030238
min         0.000000
25%         5.475000
50%        10.350000
75%        15.525000
max        21.000000
Name: tempo_relacionamento_anos, dtype: float64

faixa_tempo,clientes,tempo_medio,receita_total,receita_media_cliente,produtos_medios_cliente,meses_com_receita_medios,share_receita
0–2 anos,88,0.9397727272727273,735409.09,8356.921477272726,1.9886363636363635,9.454545454545455,0.08158471325585893
2–5 anos,135,3.4496296296296296,1164109.26,8623.031555555555,2.140740740740741,9.488888888888889,0.12914379420519556
5–10 anos,248,7.446774193548387,2333814.98,9410.544274193548,2.1653225806451615,9.818548387096774,0.25890844772605154
10–15 anos,246,12.432520325203253,2164605.58,8799.209674796748,2.130081300813008,9.56910569105691,0.2401367184029942
15+ anos,283,18.077031802120143,2616116.06,9244.226360424029,2.176678445229682,9.862190812720849,0.29022632640989987



#### ⚪ Resposta 02:

#1. Existe algum perfil de renda que atendemos melhor?
O banco atende melhor o perfil de faixa de renda até 5 mil, que representa 81% da carteira.

- 46,3% dos clientes estão nas faixa até R$ 3 mil e 35,3% entre 3 e 5 mil, 
- A receita por cliente é parecida entre as faixas, apresentando maior receita média entre >2 e <5k. 
- Os clientes que entram com a renda mais baixa de exatamente 2000 tem menor receita média.
- Em média os clientes contratam 2 produtos, independente da faixa, com leve aumento nas faixas entre 2 a 5 mil com maior volume de clientes.

#2. O tempo de relacionamento implica em resultados melhores? 
Sim para clientes na faixa de tempo de relacionamento entre 5 a 10 anos a receita média é 6% maior e na faixa de 15+ anos é 4% maior.
- Após os primeiros dois anos, a quantidade média de produtos contratados permanece próxima em 2,2 produtos.

- Porém, os clientes de 10 a 15 anos apresentam queda na receita e em quantidade de produtos, seria necessário entender se essa faixa de clientes está recebendo a mesma atenção e comunicação em relação aos outros clientes.

#3. Estamos rentabilizando nossos clientes à medida que eles permanecem na base?
Levemente sim, clientes com maior tempo de relacionamento tem maior receita e tem leve aumento de produtos contratados, de 2 para 2,15, 8% maior em relação a novos clientes.

- Considerando o potencial de rentabilização e cross sell ao longo do ciclo de vida dos clientes em relação aos novos clientes, os clientes de longa data poderiam ser melhor rentabilizados.


---


## ✅ Pergunta 3

Entendido o comportamento dos nossos produtos, nosso Head de Negócios gostaria de focar em uma estratégia de abertura de conta através da oferta de produtos mais bem direcionados. 

- Quais os 3 principais produtos mais contratados pós abertura de conta? 
- O perfil digital do cliente ou a região do cliente podem mudar nossa estratégia?


#### ⏳ Desenvolvimento 03:

In [0]:
#Pergunta 3
# Quais os 3 principais produtos mais contratados pós abertura de conta?
# Cartão 86%, crédito 7% e 5% invest

#groupby por cliente e prod id verificando menor anomes
produto_primeiro_mes = (
    db1.groupby(['cliente_id', 'produto_id'], as_index=False)
    .agg(primeiro_mes_produto=('ano_mes', 'min'))
)

#groupby por cliente id agregando o primeiro produto contratado
primeiro_mes_cliente = (
    produto_primeiro_mes.groupby('cliente_id', as_index=False)
    .agg(primeiro_mes_cliente=('primeiro_mes_produto', 'min'))
)

# criando a dtb com os dados de primeiro produto por cliente
jornada_produto = (
    produto_primeiro_mes
    .merge(primeiro_mes_cliente, on='cliente_id', how='left')
    .merge(clientes, on='cliente_id', how='left')
)

# marcar quais produtos foram contratados no primeiro mês do cliente
jornada_produto['produto_primeira'] = jornada_produto['primeiro_mes_produto'].eq(jornada_produto['primeiro_mes_cliente'])

#filtrar somente os produtos de primeira compra
db_primeira_compra = (
    jornada_produto[jornada_produto['produto_primeira']]
    .groupby('produto_id')
    .agg(clientes=('cliente_id', 'nunique'))
    .reset_index()
    .sort_values('clientes', ascending=False)
)
#calcular % sobre a base total
db_primeira_compra['share_base'] = db_primeira_compra['clientes'] / clientes_total

display(db_primeira_compra)

produto_id,clientes,share_base
CARTAO,852,0.852
CREDITO,75,0.075
INVEST,51,0.051
SEGURO,36,0.036
CONSORCIO,11,0.011


In [0]:
# O perfil digital do cliente ou a região do cliente podem mudar nossa estratégia?
# Não, o cartão continua dominando mesmo em outras faixas de perfil digital.

display(clientes['perfil_digital'].value_counts())
display(clientes['regiao'].value_counts())
    # principalmente o perfil de clientes é digital alto e do sudeste.


# produtos por perfil digital buscando a base de primeira compra
perfil_digital = jornada_produto[jornada_produto['produto_primeira']].copy()

# agrupando a base por perfil digital e produto id, contando por cliente id
perfil_digital = (
    perfil_digital.groupby(['perfil_digital', 'produto_id'])
    .agg(clientes=('cliente_id', 'nunique'))
    .reset_index()
) 
# groupby na clientes por perfil digital para ter o total por cliente
base_digital = clientes.groupby('perfil_digital').agg(clientes_segmento=('cliente_id', 'nunique')).reset_index()

# merge entre a base de primeira compra e perfil digital
perfil_digital = perfil_digital.merge(base_digital, on='perfil_digital')

# criacao de uma coluna com a porcentagem de clientes por perfil
perfil_digital['penetracao_segmento'] = perfil_digital['clientes'] / perfil_digital['clientes_segmento']

# sort
perfil_digital = perfil_digital.sort_values(['perfil_digital', 'clientes'], ascending=[True, False])

display(perfil_digital)


ALTO     667
MEDIO    177
BAIXO    156
Name: perfil_digital, dtype: int64

SUDESTE         449
NORDESTE        204
SUL             181
NORTE            92
CENTRO-OESTE     74
Name: regiao, dtype: int64

perfil_digital,produto_id,clientes,clientes_segmento,penetracao_segmento
ALTO,CARTAO,583,667,0.8740629685157422
ALTO,CREDITO,46,667,0.06896551724137931
ALTO,INVEST,30,667,0.044977511244377814
ALTO,SEGURO,20,667,0.029985007496251874
ALTO,CONSORCIO,6,667,0.008995502248875561
BAIXO,CARTAO,124,156,0.7948717948717948
BAIXO,CREDITO,16,156,0.10256410256410256
BAIXO,INVEST,11,156,0.07051282051282051
BAIXO,SEGURO,5,156,0.03205128205128205
BAIXO,CONSORCIO,2,156,0.01282051282051282


Databricks visualization. Run in Databricks to view.

Databricks visualization. Run in Databricks to view.

In [0]:
# produtos por região, buscando a base de primeira compra
perfil_regiao = jornada_produto[jornada_produto['produto_primeira']].copy()

# agrupando a base por regiao e produto id, contando por cliente id
perfil_regiao = (
    perfil_regiao.groupby(['regiao', 'produto_id'])
    .agg(clientes=('cliente_id', 'nunique'))
    .reset_index()
)

# groupby na clientes por regiao para ter o total por cliente
base_regiao = clientes.groupby('regiao').agg(clientes_segmento=('cliente_id', 'nunique')).reset_index()

# merge entre a base de primeira compra e regiao
perfil_regiao = perfil_regiao.merge(base_regiao, on='regiao')

# criacao de uma coluna com a porcentagem de clientes por regiao
perfil_regiao['penetracao_segmento'] = perfil_regiao['clientes'] / perfil_regiao['clientes_segmento']

# sort
perfil_regiao = perfil_regiao.sort_values(['regiao', 'clientes'], ascending=[True, False])

display(perfil_regiao)

regiao,produto_id,clientes,clientes_segmento,penetracao_segmento
CENTRO-OESTE,CARTAO,63,74,0.8513513513513513
CENTRO-OESTE,CREDITO,6,74,0.08108108108108109
CENTRO-OESTE,CONSORCIO,3,74,0.04054054054054054
CENTRO-OESTE,INVEST,3,74,0.04054054054054054
CENTRO-OESTE,SEGURO,1,74,0.013513513513513514
NORDESTE,CARTAO,180,204,0.8823529411764706
NORDESTE,CREDITO,13,204,0.06372549019607843
NORDESTE,INVEST,11,204,0.05392156862745098
NORDESTE,SEGURO,3,204,0.014705882352941176
NORDESTE,CONSORCIO,2,204,0.00980392156862745


Databricks visualization. Run in Databricks to view.

Databricks visualization. Run in Databricks to view.


#### ⚪ Resposta 03:
Quais os 3 principais produtos mais contratados pós abertura de conta?
- Os 3 principais produtos são:
> - Cartão com 852 clientes, 85,2% da base
> - Crédito com 75 clientes, 7,5% da base
> - Investimentos com 51 clientes, 5,1% da base

    > Cartão é o produto de entrada principal, crédito e investimentos aparecem como próximos produtos após abertura e ativação do cartão.
    > Seguro e consórcio tem menor penetração, mas são produtos menores e mais recentes, talvez com menor força de aquisição em relação ao investimento e crédito.

O perfil digital do cliente ou a região do cliente podem mudar nossa estratégia?
- Sim, para o público de perfil digital baixo a estratégia pode ser oferecer o produto cartão com o crédito como cross-sell, com abordagem mais assistida e simplificada no digital.
- Para o perfil médio, o seguro ganha relevância e o grupo pode ser tratado como grupo de transição, também seguindo a linha de abordagem assistida e simplificada para melhorar o uso digital.
- Para o perfil digital alto a estrategia pode ser oferecer o produto cartão aliado com investimentos e crédito pelo canal digital.

A análise por região mostra que sim, a região pode mudar a estratégia de produtos auxiliares, porém, cartão continua seguindo como produto principal, é possível ajustar a oferta conforme perfil digital e região.

- Sudeste tem a maior concentração dos clientes, relevante para escalar os produtos auxiliares, porém com menor volume de consorcio.
- Centro Oeste apresenta maior volume de consorcio entre todas as regiões, a estratégia pode ser oferecer cartão + crédito e consorcio.
- Nordeste cartão + crédito e investimentos
- Sul apresenta a menor concentração de cartoes e maior concentração de seguros, investimentos e crédito.
- Norte tem maior volume de cartões seguido por crédito.

---


## 🚀 Pergunta 4
O banco está desenvolvendo uma estratégia de cross-sell: aumentar a quantidade média de produtos contratados por cliente.
Hoje, sabemos quais produtos cada cliente já possui, mas queremos antecipar qual seria o próximo produto mais provável de ser adquirido, para poder direcionar campanhas de marketing e ofertas personalizadas.

Questão de negócio
> O Diretor de Marketing nos perguntou: Dado o histórico de contratação dos clientes, quais produtos geralmente são contratados depois de outro?
> Com base nesse padrão, se um cliente possui atualmente apenas alguns produtos, qual produto deveríamos recomendar como próximo passo natural da jornada dele?
> 

#### ⏳ Desenvolvimento 04:

In [0]:
# para cada cliente, identificar o primeiro mes que o produto apareceu, como já feito na parte de primeiro produto, e em seguida encontrar os proximos produtos que aparecem em seguida.

# ajustar o ano mes

if not pd.api.types.is_datetime64_any_dtype(db1['ano_mes']):
    db1['ano_mes_dt'] = pd.to_datetime(db1['ano_mes'] + '-01')
else:
    db1['ano_mes_dt'] = db1['ano_mes']


# Garantir ordenação correta
produto_primeiro_mes = produto_primeiro_mes.sort_values(
    ['cliente_id', 'primeiro_mes_produto']
)

# Criar pares origem - destino
pares = []
for cliente_id, grupo in produto_primeiro_mes.groupby('cliente_id'):
    grupo = grupo.sort_values('primeiro_mes_produto')
    
    for i, origem in grupo.iterrows():
        for j, destino in grupo.iterrows():
            
            # Só considera destino se veio depois da origem
            if destino['primeiro_mes_produto'] > origem['primeiro_mes_produto']:
                pares.append({
                    'cliente_id': cliente_id,
                    'produto_origem': origem['produto_id'],
                    'produto_destino': destino['produto_id'],
                    'mes_origem': origem['primeiro_mes_produto'],
                    'mes_destino': destino['primeiro_mes_produto']
                })

pares_produtos = pd.DataFrame(pares)

display(pares_produtos.head(5))

cliente_id,produto_origem,produto_destino,mes_origem,mes_destino
100000,CARTAO,SEGURO,2024-04-01T00:00:00.000Z,2024-10-01T00:00:00.000Z
100001,CARTAO,SEGURO,2024-04-01T00:00:00.000Z,2024-08-01T00:00:00.000Z
100002,CARTAO,SEGURO,2024-04-01T00:00:00.000Z,2024-11-01T00:00:00.000Z
100004,CARTAO,CREDITO,2024-02-01T00:00:00.000Z,2024-08-01T00:00:00.000Z
100004,CARTAO,INVEST,2024-02-01T00:00:00.000Z,2024-09-01T00:00:00.000Z


In [0]:
# verificar os pares ex cartao/seguros, seguros/cartao, cartao/invest etc.
# groupby na pares_produtos trazendo clientes unicos por prod origem e prod destino, com sort asc 
resumo_pares = (
    pares_produtos
    .groupby(['produto_origem', 'produto_destino'], as_index=False)
    .agg(clientes=('cliente_id', 'nunique'))
    .sort_values('clientes', ascending=False)
)

display(resumo_pares)

# Quais produtos geralmente são contratados depois de outro?
# O produto mais contratado é o cartão, os produtos mais contratados com cartão como origem -> são cartão -> credito 389 clientes, 25% do total (389/1576), cartão -> investimento 314 clientes, 20% e cartão -> seguro 250 clientes, 16%.
# O interessante dessa visualização é perceber que crédito -> investimentos 96 clientes 6% e investimentos 92 clientes 6% -> crédito possuem boa relação quando sao produtos de origem, fortalecendo a ideia do cross-sell entre eles, além do cartão.

produto_origem,produto_destino,clientes
CARTAO,CREDITO,389
CARTAO,INVEST,314
CARTAO,SEGURO,250
CREDITO,INVEST,96
INVEST,CREDITO,92
CARTAO,CONSORCIO,87
CREDITO,SEGURO,66
SEGURO,CREDITO,62
INVEST,SEGURO,61
SEGURO,INVEST,38


Databricks visualization. Run in Databricks to view.


#### ⚪ Resposta 04:
> Quais produtos geralmente são contratados depois de outro?

O cartão aparece como principal produto de origem na jornada de contratação, os produtos mais contratados com cartão como origem são:
- cartão -> credito 389 clientes, 37% de cartão (389/1040) e 25% do total (389/1.576), 
- cartão -> investimento 314 clientes, 30% de cartão e 20% do total,
- cartão -> seguro 250 clientes, 24% de cartão e 16% do total.

Após os 3 principais, 
- crédito -> investimentos 96 clientes, 53% de crédito e 6% do total
- investimentos -> crédito 92 clientes, 53% de investimentos e 6% do total

Crédito e investimentos possuem boa relação quando sao produtos de origem, fortalecendo a ideia do cross-sell entre eles, além do cartão.

> Com base nesse padrão, se um cliente possui atualmente apenas alguns produtos, qual produto deveríamos recomendar como próximo passo natural da jornada dele?

Com base apenas nessa tabela, deveria ser recomendado na ordem: 
Cartão: Crédito > investimento > seguro > consorcio
Crédito: investimentos > seguros > consorcio > cartão
Investimento: Crédito > seguro > consorcio > cartão
Consórcio: Crédito > investimento > seguro > cartão
Seguros: Crédito > investimento > consorcio > cartão

Cartão aparece por ultimo por ser o produto de entrada, já que não foi o produto pelo qual o cliente buscou o banco, seguindo essa lógica é possível que não seja o melhor produto para cross-sell.

Com mais tempo de estudo e análise, seria possível aprofundar mais usando recursos de estatistica, segmentacoes por perfil de cliente, renda, tempo de relacionamento e receita gerada para definir com mais certeza as possibilidades de cross-sell.

In [0]:
# Olá, formulei as ideias do que queria seguir e usei ia (gpt5.5 e o genie do databricks) para a escrita dos codigos, ultimamente trabalhei mais com power query do que sql/python, acredito que o principal seja a ideia para chegar nos resultados.
# Priorizei usar a ia como ferramenta, como as ferramentas de point and click, ao inves de montar um groupby/merge na mão. 

# Obrigado pelo tempo!

---